Jules.md

```diff
@@ -0,0 +1,553 @@
+    services.
+*   **Operation-Specific Options**: Some operations might take optional
+    parameters to control behavior, though not always through a dedicated
+    "Options" class like Firestore's `SetOptions`. For example,
+    `User::UpdateUserProfile()` takes a `UserProfile` struct.
+
+### Listeners for Real-time Updates
+
+Many Firebase services support real-time data synchronization using listeners.
+
+*   **`firebase::database::ValueListener` /
+    `firebase::database::ChildListener`**: Implemented by the developer and
+    registered with a `DatabaseReference`.
+    *   `ValueListener::OnValueChanged(const firebase::database::DataSnapshot& snapshot)`
+        is called when the data at that location changes.
+    *   `ChildListener` has methods like `OnChildAdded()`, `OnChildChanged()`,
+        `OnChildRemoved()`.
+*   **`firebase::auth::AuthStateListener`**: Implemented by the developer and
+    registered with `firebase::auth::Auth`.
+    *   `AuthStateListener::OnAuthStateChanged(firebase::auth::Auth* auth)` is
+        called when the user's sign-in state changes.
+*   **Removing Listeners**: Listeners are typically removed by passing the
+    listener instance to a corresponding `Remove...Listener()` method (e.g.,
+    `reference->RemoveValueListener(my_listener);`,
+    `auth->RemoveAuthStateListener(my_auth_listener);`).
+
+This overview provides a general understanding. Always refer to the specific
+header files in `firebase/app/client/cpp/include/firebase/` and
+`firebase/product_name/client/cpp/include/firebase/product_name/` for detailed
+API documentation.
+
+# Best Practices
+
+## Coding Style
+
+*   **Google C++ Style Guide**: Adhere to the
+    [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html)
+    as mentioned in `CONTRIBUTING.md`.
+*   **Formatting**: Use `python3 scripts/format_code.py -git_diff -verbose` to
+    format your code before committing.
+
+## Comments
+
+*   Write clear and concise comments where necessary to explain complex logic or
+    non-obvious behavior.
+*   **Avoid overly verbose comments**: Do not state the obvious. The code should
+    be as self-documenting as possible.
+*   Follow existing comment styles within the module you are working on.
+*   "Avoid adding comments next to `#include` directives merely to explain why
+    the include is necessary; the code usage should make this clear, or it can
+    be part of a broader comment block if truly non-obvious for a section of
+    code."
+*   "Do not include comments that narrate the AI agent's iterative development
+    process (e.g., 'Removed old logic here', 'Changed variable name from X to
+    Y because...', 'Attempted Z but it did not work'). Comments should focus on
+    explaining the current state of the code for future maintainers, not its
+    development history or the AI's thought process."
+
+## Error Handling
+
+*   **Check `Future` status and errors**: Always check `future.status()` and
```

Comment:
Should we have a thing about making sure that Futures that are returned are somehow completed / not abandoned?